woocommerce/woocommerce-ios
# Domain selector for paid domains: UI integration

Part of #8558 

## Description


For a user to choose a paid domain from WPCOM, I wanted to reuse the pre-existing `DomainSelector` classes/structs for the free domain selector in the store creation flow. However, a challenge surfaced since the schema of a free domain and paid domain is quite different. For a free domain, we just need to know the domain name. In order to purchase a paid domain, we need to know other info about the domain like the WPCOM product ID. It doesn't make sense to create a protocol for different domain suggestion types, since they are so different. I decided to make the domain suggestion type generic with a protocol for UI display `DomainSuggestionViewProperties`, and add a protocol `DomainSelectorDataProvider` for an implementation to return a generic domain suggestion type. There are two implementations, `FreeDomainSelectorDataProvider` that returns `[FreeDomainSuggestionViewModel]` and `PaidDomainSelectorDataProvider` that returns `[PaidDomainSuggestionViewModel]`. `FreeDomainSuggestionViewModel`/`PaidDomainSuggestionViewModel` can then be used to display a domain row.

To integrate the paid domain selector into the domain settings flow, I created a coordinator `DomainSettingsCoordinator` to navigate to the paid domain selector hosting controller `PaidDomainSelectorHostingController` when the user taps to add a domain. Sorry for the larger size of the PR than I hoped, it's harder to test the integration otherwise 🙇🏻‍♀️

Please note that there are a few other UI subtasks https://github.com/woocommerce/woocommerce-ios/issues/8558 for this screen:
- Attributed price info
- Don't show support action in the navigation bar
- Update header text

## Testing instructions


Prerequisite: a WC store with a WPCOM plan is required

- Log in if needed, and continue with a WPCOM store
- Go to the Menu tab
- Tap on the settings CTA
- Tap the `Domain` row
- Tap `Add a domain` --> domain selector with paid domains should be shown, and each domain should have price info
- Feel free to search for other domains --> each domain should have price info
- Select a domain and tap `Continue` --> a message about the selected domain should be shown in the console

---
- [x] @jaclync tests that the free domain selector works as before for store creation

## Screenshots


paid domains | free domains
-- | --
![Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-17 at 15 27 41](https://user-images.githubusercontent.com/1945542/212835665-0c0c2e7c-ecb7-4468-abc8-93abee7e9ea2.png) | ![Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-17 at 15 29 05](https://user-images.githubusercontent.com/1945542/212835727-7d274083-fba3-4aac-97d8-f62e731b53f9.png)


---
